packages/babel-helpers/src/helpers.js

```diff
@@ -1973,3 +1973,122 @@ helpers.wrapRegExp = helper("7.2.6")`
     return _wrapRegExp.apply(this, arguments);
   }
 `;
+helpers.slice = helper("7.6.0")`
+  import isNativeFunction from 'isNativeFunction';
+  function isString(value) {
+    return typeof value === "string" || (
+        typeof value === "object" && Object.prototype.toString.call(value) === "[object String]"
+    )
+  }
+
+  function toLength(value) {
+    var len = toInteger(value);
+    if (len < 0) {
+      return 0;
+    }
+    return Math.min(len, 9007199254740991);
+  }
+
+  function toInteger(value) {
+    const number = +value;
+    if (isNaN(number)) {
+      return 0;
+    }
+    if (number === 0 || number === Infinity || number === -Infinity) {
+      return number;
+    }
+    return number > 0 ? Math.floor(number) : -Math.floor(-number);
+  }
+
+  export default function slice(t, begin, end, step) {
+    // The default value for the step argument is 1
+    if (step === void 0) {
+      step = 1;
+    }
+    step = toInteger(step);
+    if (step === 0 || step === Infinity || step === -Infinity) {
+      // todo: the proposal does not specify details about these cases
+      // https://github.com/tc39/proposal-slice-notation/issues/28
+      throw new Error("Slice step can not be zero, NaN of Infinity");
+    }
+    var tIsString = isString(t);
+    var tIsArray = Array.isArray(t);
+    if (
+        !tIsArray && !tIsString &&
+        (
+          typeof t.slice !== "function" ||
+          !isNativeFunction(t.slice) ||
+          Object.prototype.toString.call(t).indexOf("ArrayBuffer") !== -1
+        )
+    ) {
+      // todo: may desugar to a [Symbol.slice] method call for custom objects
+      // https://github.com/tc39/proposal-slice-notation/issues/1
+      throw new Error("Slice notation only supports array, string and typed array");
+    }
+    // fast path for step = 1
+    if (step === 1) {
+      return t.slice(begin, end);
+    }
+    var k, final, low, high;
+    var len = toLength(t.length);
+    if (step < 0) {
+      // todo: the proposal does not specify inclusivity on negative step
+      // here we align behaviour to Python slice expression
+      // https://github.com/tc39/proposal-slice-notation/issues/29
+      low = -1;
+      high = len - 1;
+    } else {
+      low = 0;
+      high = len;
+    }
+    if (begin === void 0) {
+      k = step < 0 ? high : low;
+    } else {
+      begin = toInteger(begin);
+      k = begin < 0 ? Math.max(low, len + begin) : Math.min(begin, high);
+    }
+    if (end === void 0) {
+      final = step < 0 ? low : high;
+    } else {
+      end = toInteger(end);
+      final = end < 0 ? Math.max(low, len + end) : Math.min(end, high);
+    }
+
+    var o;
+    if (tIsString) {
+      o = "";
```

Comment:
Nit: We can use the array path for this, and join at the end.

Reply:
My major concern on implementation is memory footprint and performance. We could shorten the helper if we convert string to array, but that means given a string `str` with length `N`, `str[::M]` will require `N` byte memory while the optimal should be `N/M`.